superset/db_engine_specs/hana.py

```diff
@@ -0,0 +1,39 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from datetime import datetime
+
+from superset.db_engine_specs.base import LimitMethod
+from superset.db_engine_specs.postgres import PostgresBaseEngineSpec
+
+
+class HanaEngineSpec(PostgresBaseEngineSpec):
+    engine = "hana"
+    limit_method = LimitMethod.WRAP_SQL
+    force_column_alias_quotes = True
+    max_column_name_length = 30
+
+    _time_grain_functions = {
+        None: '{col}',
+        'P1D': "TO_DATE({col})",
+        'P1M': "TO_DATE(SUBSTRING(to_date({col}),0,7)||'-01')",
+        'P0.25Y': "TO_DATE(SUBSTRING(to_date({col}),0,5)||'0'||SUBSTRING(QUARTER(TO_DATE({col}), 1),7,1)||'-01')",
```

Comment:
The Quarter calculation is wrong here.
You may want to use either of these:
```
        'P0.25Y': "SUBSTRING(to_date({col}), 0, 5)|| lpad(cast(FLOOR((extract(month from {col})-1)/3)*3+1 as text),2,'0') ||'-01'",
```
or
```
        'P0.25Y': "SUBSTRING(to_date({col}), 0, 5)|| lpad(cast((cast(SUBSTRING(QUARTER(TO_DATE({col}), 1), 7, 1) as int)-1)*3 +1 as text),2,'0') ||'-01'",
```
Not sure if you need to wrap it in another `to_date()` like you did, though. For me it works without it.

Background:  
HANA's `QUARTER()` function yields something like, e.g., 2017-Q4, So we cannot just use the last digit as month but need to convert it to the first month of the quarter.  
Furthermore, we need to use `LPAD()` instead of just prepending a `0` to the extracted month, since otherwise Q4 (October) would end up as `010` instead of `10`.

Reply:
![image](https://user-images.githubusercontent.com/44701753/67271613-6a56a300-f4ed-11e9-8cc6-428dfc8fbdbe.png)
What I need is the natural season. Your needs may not be the same as mine

Reply:
It is not at all about anyone's personal needs, I guess, but it may be about a misunderstanding on either side of us:

In my understanding the `P0.25Y` time granularity is supposed to yield the date of the first month of the corresponding quarter. So, e.g., `2019-Q2` needs to be translated to `2019-04-01`. However, your code translates it to `2019-02-01` which is simply wrong here.

You may want to play with this development snippet; it may help you understand my point of view:

```
select
	sub.dt
	,QUARTER(TO_DATE(sub.dt), 1) as q_date
	,SUBSTRING(QUARTER(TO_DATE(sub.dt), 1), 7, 1) as q_num
	,cast(cast(SUBSTRING(QUARTER(TO_DATE(sub.dt), 1), 7, 1) as int) *3 -2 as text) as month_num
	,cast(FLOOR((extract(month from sub.dt)-1)/3)*3+1 as text) as month_num2
	,SUBSTRING(to_date(sub.dt), 0, 5)||'0'||SUBSTRING(QUARTER(TO_DATE(sub.dt), 1), 7, 1)||'-01' AS wrong_quarter_start_date
	,SUBSTRING(to_date(sub.dt), 0, 5)|| lpad(cast((cast(SUBSTRING(QUARTER(TO_DATE(sub.dt), 1), 7, 1) as int)-1)*3 +1 as text),2,'0') ||'-01' AS correct_quarter_start_date
	,SUBSTRING(to_date(sub.dt), 0, 5)|| lpad(cast(FLOOR((extract(month from sub.dt)-1)/3)*3+1 as text),2,'0') ||'-01' AS correct_quarter_start_date2
from (
	select
		to_date('2017-12-28T00:00:00') as dt 
	from dummy
) sub
```

Reply:
BTW:  
My understanding of what this time grain is supposed to yield is based on how it is implemented in Superset's the rather well established DB specification of PostgreSQL:

![image](https://user-images.githubusercontent.com/1220356/67272581-1c29ab80-f4bd-11e9-87f6-6fd3ed5e9497.png)

Reply:
You're right. I ignored the front end